# EventPulse AI
## Dataset Analysis & Feature Engineering

### Flipkart Grid 7.0 - Gridlock 2.0

Objective:
Analyze historical event data to identify traffic patterns, engineer predictive features, and develop a foundation for event-driven congestion forecasting and resource recommendation.

###Import Libraries

In [87]:
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns

from datetime import datetime

pd.set_option("display.max_columns", None)

print("Libraries Loaded Successfully")

Libraries Loaded Successfully


##Phase 1: Load Dataset

In [88]:
df = pd.read_csv("/content/Astram event data_anonymized - Astram event data_anonymizedb40ac87.csv")
print("Shape:", df.shape)
df.head()

Shape: (8173, 46)


,id,event_type,latitude,longitude,endlatitude,endlongitude,address,end_address,event_cause,requires_road_closure,start_datetime,end_datetime,status,authenticated,modified_datetime,map_file,direction,description,veh_type,veh_no,corridor,priority,cargo_material,reason_breakdown,age_of_truck,created_date,route_path,client_id,created_by_id,last_modified_by_id,assigned_to_police_id,citizen_accident_id,comment,police_station,meta_data,kgid,resolved_at_address,resolved_at_latitude,resolved_at_longitude,closed_by_id,closed_datetime,resolved_by_id,resolved_datetime,gba_identifier,zone,junction
0,FKID000000,unplanned,13.040004,77.518099,0.000000,0.000000,"Mumbai Bengaluru Highway, Jalahalli Cross Junc...",NaN,vehicle_breakdown,False,2024-03-07 17:01:48.111+00,NaN,closed,yes,2024-03-07 19:35:47.871698+00,NaN,NaN,s m circle in coming man track,lcv,FKN00GL0000,Tumkur Road,High,NaN,NaN,NaN,2024-03-07 17:03:51.164032+00,NaN,1,FKUSR00000,FKUSR00001,NaN,NaN,NaN,Peenya,NaN,FKKG000000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,FKID000001,unplanned,12.921876,77.645158,0.000000,0.000000,"19th Main Road, Heavie Halcyon, Agara, HSR Lay...",NaN,vehicle_breakdown,False,2024-01-30 04:07:24.173+00,NaN,resolved,yes,2024-01-30 04:17:46.828979+00,NaN,NaN,Starting problem,heavy_vehicle,FKN00GL0001,ORR East 1,High,NaN,NaN,NaN,2024-01-30 04:08:22.954979+00,NaN,1,FKUSR00002,FKUSR00001,NaN,NaN,NaN,HSR Layout,NaN,FKKG000001,"19th Main Road, Heavie Halcyon, Agara, HSR Lay...",12.921876,77.645158,NaN,NaN,FKUSR00002,2024-01-30 04:17:46.828355+00,NaN,NaN,NaN
2,FKID000002,unplanned,12.955622,77.585708,0.000000,0.000000,"Lalbagh Main Road, Dr Sri Shantaveera Swami Ci...",NaN,others,False,2023-11-11 06:18:03.343+00,NaN,closed,yes,2024-01-30 04:56:03.282003+00,NaN,NaN,ಊರ್ವಶಿ ಜಂಕ್ಷನ್ ನಲ್ಲಿ ಒಳಚರಂಡಿ ಚೇಂಬರ್ ಗೆ ಹೊಸದಾಗಿ...,NaN,NaN,Non-corridor,Low,NaN,NaN,NaN,2023-11-11 06:20:00.989398+00,NaN,1,FKUSR00003,FKUSR00001,NaN,NaN,NaN,Wilson Garden,NaN,FKKG000002,NaN,NaN,NaN,FKUSR00003,2024-01-30 04:56:03.281509+00,NaN,NaN,Bengaluru Central Corporation,Central Zone 2,UrvashiJunction
3,FKID000003,unplanned,13.006147,77.579435,13.006239,77.579516,"Sankey Road, Bashyam Circle, Sadashiva Nagar, ...","Sankey Road, Palace Orchard Upper, Sadashiva N...",tree_fall,True,2024-03-07 17:56:55.061+00,NaN,closed,yes,2024-03-14 07:42:05.55005+00,NaN,NaN,tree fall,NaN,NaN,Non-corridor,Low,NaN,NaN,NaN,2024-03-07 17:58:56.696892+00,NaN,1,FKUSR00004,FKUSR00001,NaN,NaN,NaN,Sadashivanagar,NaN,FKKG000003,NaN,NaN,NaN,FKUSR00004,2024-03-14 07:42:05.54944+00,NaN,NaN,NaN,NaN,NaN
4,FKID000004,unplanned,12.953980,77.585233,0.000000,0.000000,"Lalbagh Fort Road, Lalbagh Main Gate Junction,...",NaN,vehicle_breakdown,False,2024-01-30 04:56:32.348+00,NaN,closed,yes,2024-01-30 05:35:17.33908+00,NaN,NaN,[LOCATION] ಪೈಪ್ [PERSON] ವಾಹನ ಆಫ್ ಆಗಿರುತ್ತದೆ ಸರ್,private_bus,FKN00GL0002,Non-corridor,Low,NaN,NaN,NaN,2024-01-30 04:58:55.937662+00,NaN,1,FKUSR00003,FKUSR00001,NaN,NaN,NaN,Wilson Garden,NaN,FKKG000002,NaN,NaN,NaN,FKUSR00003,2024-01-30 05:35:17.338283+00,NaN,NaN,NaN,NaN,LalbaghMainGateJunc


###Dataset Overview

In [89]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8173 entries, 0 to 8172
Data columns (total 46 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   id                     8173 non-null   object 
 1   event_type             8173 non-null   object 
 2   latitude               8173 non-null   float64
 3   longitude              8173 non-null   float64
 4   endlatitude            8004 non-null   float64
 5   endlongitude           8004 non-null   float64
 6   address                8170 non-null   object 
 7   end_address            687 non-null    object 
 8   event_cause            8173 non-null   object 
 9   requires_road_closure  8173 non-null   bool   
 10  start_datetime         8173 non-null   object 
 11  end_datetime           490 non-null    object 
 12  status                 8173 non-null   object 
 13  authenticated          8173 non-null   object 
 14  modified_datetime      8173 non-null   object 
 15  map_

In [90]:
df.shape

(8173, 46)

In [91]:
df.columns.tolist()

['id',
 'event_type',
 'latitude',
 'longitude',
 'endlatitude',
 'endlongitude',
 'address',
 'end_address',
 'event_cause',
 'requires_road_closure',
 'start_datetime',
 'end_datetime',
 'status',
 'authenticated',
 'modified_datetime',
 'map_file',
 'direction',
 'description',
 'veh_type',
 'veh_no',
 'corridor',
 'priority',
 'cargo_material',
 'reason_breakdown',
 'age_of_truck',
 'created_date',
 'route_path',
 'client_id',
 'created_by_id',
 'last_modified_by_id',
 'assigned_to_police_id',
 'citizen_accident_id',
 'comment',
 'police_station',
 'meta_data',
 'kgid',
 'resolved_at_address',
 'resolved_at_latitude',
 'resolved_at_longitude',
 'closed_by_id',
 'closed_datetime',
 'resolved_by_id',
 'resolved_datetime',
 'gba_identifier',
 'zone',
 'junction']

## Observation

- The dataset contains **8,173 event records** and **46 attributes**.
- The dataset includes a combination of geospatial, temporal, operational, and event-specific information.
- Key attributes available for analysis include:
  - Event Type (`event_type`)
  - Event Cause (`event_cause`)
  - Latitude & Longitude (`latitude`, `longitude`)
  - Event Start Time (`start_datetime`)
  - Event Priority (`priority`)
  - Police Station (`police_station`)
  - Zone (`zone`)
  - Junction (`junction`)
  - Road Closure Requirement (`requires_road_closure`)
- Most columns are categorical (`object` datatype), while location coordinates and vehicle-related attributes are numerical.
- The dataset occupies approximately **2.8 MB** of memory and is suitable for rapid exploratory analysis and model development.

---

## Insight

- The dataset provides sufficient information to support both **event impact forecasting** and **traffic resource recommendation** use cases.
- The presence of geospatial attributes enables hotspot identification and map-based visualizations.
- Temporal attributes can be transformed into meaningful predictive features such as:
  - Event Duration
  - Hour of Day
  - Day of Week
  - Weekend Indicator
  - Peak Hour Indicator
- Operational attributes such as `priority` and `requires_road_closure` are expected to be strong indicators of traffic disruption severity.

---

## Action

- Perform missing value assessment to identify usable and non-usable attributes.
- Convert all datetime columns into proper datetime format.
- Categorize attributes into:
  - Spatial Features
  - Temporal Features
  - Event Features
  - Operational Features
- Identify potential target variables for congestion forecasting and resource recommendation.

In [92]:
missing_df = pd.DataFrame({
    "Missing Count": df.isnull().sum(),
    "Missing Percentage": round(df.isnull().sum()*100/len(df),2)
})

missing_df.sort_values(
    by="Missing Percentage",
    ascending=False
)

,Missing Count,Missing Percentage
comment,8173,100.00
map_file,8173,100.00
meta_data,8173,100.00
direction,8130,99.47
resolved_at_address,8099,99.09
resolved_at_latitude,8099,99.09
resolved_by_id,8099,99.09
resolved_datetime,8099,99.09
resolved_at_longitude,8099,99.09
assigned_to_police_id,8045,98.43


# Missing Value Analysis

## Observation

- Several columns exhibit extremely high levels of missing data.
- Three columns contain **100% missing values**:
  - `comment`
  - `map_file`
  - `meta_data`
- Several columns have more than **95% missing values**, including:
  - `direction`
  - `resolved_at_address`
  - `resolved_at_latitude`
  - `resolved_at_longitude`
  - `resolved_by_id`
  - `resolved_datetime`
  - `assigned_to_police_id`
  - `citizen_accident_id`
  - `route_path`
  - `cargo_material`
  - `reason_breakdown`
  - `age_of_truck`
- Event closure-related fields such as `closed_datetime` and `closed_by_id` contain over 60% missing values.
- Location enrichment fields such as `zone` and `junction` contain significant missing values:
  - `zone`: 57.86%
  - `junction`: 69.29%
- Core event attributes including:
  - `event_type`
  - `event_cause`
  - `latitude`
  - `longitude`
  - `start_datetime`
  - `requires_road_closure`
  - `police_station`
  - `status`
  
  are complete with no missing values.

---

## Insight

- The dataset contains a strong set of operational and event-related features despite several sparsely populated administrative columns.
- Core fields required for event forecasting remain highly reliable and can serve as the foundation for predictive modeling.
- Attributes related to event resolution and accident investigation appear to be recorded only for a small subset of incidents and are unlikely to contribute meaningful predictive value.
- High missingness in `zone` and `junction` suggests that relying heavily on these features may reduce the usable dataset size.
- Vehicle-related attributes (`veh_type`, `veh_no`) may be useful for exploratory analysis but are unsuitable as primary predictive features due to approximately 40% missing values.

---

## Action

### Drop Columns (Missing Percentage > 95%)

The following columns will be removed due to insufficient information content:

- `comment`
- `map_file`
- `meta_data`
- `direction`
- `resolved_at_address`
- `resolved_at_latitude`
- `resolved_at_longitude`
- `resolved_by_id`
- `resolved_datetime`
- `assigned_to_police_id`
- `citizen_accident_id`
- `route_path`
- `cargo_material`
- `reason_breakdown`
- `age_of_truck`

### Retain for Further Evaluation

The following columns may still provide value and will be evaluated further:

- `zone`
- `junction`
- `veh_type`
- `veh_no`
- `description`
- `closed_datetime`
- `closed_by_id`

### Preserve Core Predictive Features

The following columns will be prioritized during feature engineering:

- `event_type`
- `event_cause`
- `requires_road_closure`
- `start_datetime`
- `priority`
- `latitude`
- `longitude`
- `police_station`
- `status`

In [93]:
df.duplicated().sum()

np.int64(0)

# Duplicate Record Analysis

## Observation

- No duplicate records were found in the dataset.
- The dataset contains **8,173 unique event entries**.

---

## Insight

- The absence of duplicate records indicates that the event logging process is relatively consistent.
- No duplicate removal procedures are required, reducing the risk of introducing bias during preprocessing.
- All available records can be utilized for exploratory analysis and model development.

---

## Action

- Retain all records for further analysis.
- Proceed to event distribution analysis to understand the composition of planned and unplanned events.

##Phase 2: Exploratory Data Analysis

In [94]:
df['event_type'].value_counts()

,count
event_type,
unplanned,7706
planned,467


# Event Type Distribution Analysis

## Observation

- The dataset is highly imbalanced.
- Event distribution:

| Event Type | Count |
|------------|--------|
| Unplanned | 7,706 |
| Planned | 467 |

- Approximately **94.3%** of all events are unplanned.
- Only **5.7%** of events are planned.

---

## Insight

- Traffic management authorities primarily deal with unexpected incidents rather than pre-scheduled events.
- The dominance of unplanned events highlights the importance of rapid response and resource allocation.
- Predictive systems capable of assessing incident severity and recommending operational actions can significantly improve traffic management effectiveness.

---

## Action

- Analyze event causes to understand the major contributors to traffic disruptions.
- Investigate whether planned and unplanned events differ in:
  - Priority
  - Road Closure Requirements
  - Spatial Distribution
  - Operational Impact

In [95]:
df['event_cause'].value_counts()

,count
event_cause,
vehicle_breakdown,4896
others,638
pot_holes,537
construction,480
water_logging,458
accident,365
tree_fall,284
road_conditions,170
congestion,136


# Event Cause Analysis

## Observation

- The dataset contains a diverse set of event causes affecting traffic operations.
- The most frequently occurring causes are:

| Event Cause | Count |
|------------|--------:|
| Vehicle Breakdown | 4,896 |
| Others | 638 |
| Pot Holes | 537 |
| Construction | 480 |
| Water Logging | 458 |
| Accident | 365 |
| Tree Fall | 284 |
| Road Conditions | 170 |
| Congestion | 136 |
| Public Event | 84 |
| Procession | 72 |
| VIP Movement | 20 |
| Protest | 15 |

- Vehicle breakdowns account for the majority of recorded events.
- Planned-event-related causes such as public events, processions, protests, and VIP movements constitute a relatively small proportion of the dataset.

---

## Insight

- Traffic disruptions are primarily driven by operational incidents rather than large public gatherings.
- Vehicle breakdowns represent the most significant contributor to traffic management interventions.
- Infrastructure-related issues such as potholes, construction activities, water logging, and road conditions collectively account for a substantial portion of traffic disruptions.
- Event management systems should not focus solely on planned gatherings but should also support incident-response decision-making.

---

## Action

- Categorize event causes into broader operational groups for better modeling.
- Analyze relationships between event causes and:
  - Priority levels
  - Road closure requirements
  - Geographic distribution
- Investigate whether specific event causes consistently result in higher operational impact.

In [96]:
df["priority"].value_counts(dropna=False)

,count
priority,
High,5030
Low,3141
NaN,2


# Priority Distribution Analysis

## Observation

- The `priority` attribute contains only two operational levels:
  
| Priority Level | Count |
|---------------|--------:|
| High | 5,030 |
| Low | 3,141 |
| Missing | 2 |

- High-priority events account for approximately **61.55%** of all recorded incidents.
- Low-priority events account for approximately **38.43%** of all recorded incidents.
- Missing values are negligible (2 records).

---

## Insight

- The priority field appears to represent the urgency or operational severity of an event.
- Since priority labels already exist for nearly all records, they provide a strong candidate target variable for supervised machine learning.
- Predicting event priority could help authorities proactively assess the severity of newly reported incidents and allocate resources accordingly.
- The class distribution is reasonably balanced, reducing concerns about severe class imbalance during model training.

---

## Action

- Remove the two records containing missing priority values.
- Investigate relationships between priority and:
  - Event Cause
  - Road Closure Requirement
  - Event Type
  - Location
  - Time of Occurrence
- Evaluate the feasibility of using `priority` as the primary prediction target for the EventPulse AI forecasting engine.

In [97]:
df["requires_road_closure"].value_counts()

,count
requires_road_closure,
False,7497
True,676


# Road Closure Requirement Analysis

## Observation

- The majority of recorded events do not require road closures.

| Road Closure Required | Count |
|----------------------|--------:|
| False | 7,497 |
| True | 676 |

- Only **8.27%** of all events require road closures.
- Approximately **91.73%** of events are managed without complete road closures.

---

## Insight

- Road closures represent relatively rare but operationally significant incidents.
- Events requiring road closures are likely associated with higher traffic disruption and increased resource requirements.
- The rarity of road closure events suggests that this feature may serve as a strong indicator of event severity.

---

## Action

- Analyze the relationship between road closure requirements and event priority.
- Include `requires_road_closure` as a key predictive feature during model development.
- Evaluate whether road closure events exhibit distinct spatial or temporal patterns.

In [98]:
pd.crosstab(
    df["event_cause"],
    df["priority"]
)

priority,High,Low
event_cause,,
Debris,8,4
Fog / Low Visibility,1,1
accident,168,197
congestion,94,42
construction,302,178
debris,1,0
others,380,258
pot_holes,299,238
procession,23,49


# Event Cause vs Priority Analysis

## Observation

- Event causes exhibit varying distributions across priority levels.

### High Priority Dominant Causes

| Event Cause | High | Low |
|------------|------:|------:|
| Vehicle Breakdown | 3241 | 1653 |
| Construction | 302 | 178 |
| Water Logging | 271 | 187 |
| Congestion | 94 | 42 |

### Low Priority Dominant Causes

| Event Cause | High | Low |
|------------|------:|------:|
| Tree Fall | 93 | 191 |
| Procession | 23 | 49 |
| VIP Movement | 7 | 13 |
| Protest | 6 | 9 |

### Nearly Balanced Causes

| Event Cause | High | Low |
|------------|------:|------:|
| Public Event | 42 | 42 |

---

## Insight

- Event cause appears to be one of the strongest determinants of operational priority.
- Vehicle breakdowns account for the largest share of high-priority incidents.
- Infrastructure-related issues such as construction and water logging are frequently classified as high priority.
- Public events show a balanced distribution, suggesting that additional contextual factors may influence their operational severity.
- Certain event types consistently trend toward lower operational priority.

---

## Action

- Treat `event_cause` as a primary predictive feature.
- Create higher-level event categories such as:
  - Incident
  - Infrastructure
  - Public Event
  - Weather
- Evaluate feature importance during model training to quantify the contribution of event causes to priority prediction.

In [99]:
pd.crosstab(
    df["requires_road_closure"],
    df["priority"],
    margins=True
)

priority,High,Low,All
requires_road_closure,,,
False,4733,2762,7495
True,297,379,676
All,5030,3141,8171


# Road Closure vs Priority Analysis

## Observation

| Road Closure Required | High Priority | Low Priority |
|----------------------|--------------:|-------------:|
| False | 4733 | 2762 |
| True | 297 | 379 |

- Most recorded events do not require road closures.
- Events requiring road closures account for only a small fraction of the dataset.
- Among road-closure events, low-priority incidents slightly outnumber high-priority incidents.

---

## Insight

- Road closure status alone is not a strong indicator of operational priority.
- Contrary to initial expectations, events involving road closures are not automatically classified as high priority.
- Traffic authorities likely consider multiple factors such as event cause, location, expected disruption, and resource requirements while assigning priorities.
- This suggests that event severity is influenced by a combination of contextual factors rather than a single operational attribute.

---

## Action

- Retain `requires_road_closure` as a predictive feature but avoid treating it as a primary indicator of severity.
- Investigate interactions between road closure requirements and:
  - Event Cause
  - Event Type
  - Location
  - Time of Occurrence
- Build a multi-factor prediction model that incorporates both operational and contextual event characteristics.
- Proceed with temporal analysis to understand how event occurrence varies across different time periods.

In [100]:
df['start_datetime'] = pd.to_datetime(
    df['start_datetime'],
    format='mixed',
    errors='coerce'
)

In [101]:
df['start_datetime'].isnull().sum()

np.int64(0)

In [102]:
df['hour'] = df['start_datetime'].dt.hour
df['day_of_week'] = df['start_datetime'].dt.day_name()
df['month'] = df['start_datetime'].dt.month_name()

In [103]:
df['hour'].value_counts().sort_index()

,count
hour,
0,421
1,384
2,395
3,374
4,560
5,661
6,662
7,482
8,330


# Hourly Event Distribution Analysis

## Observation

- Event occurrences are not uniformly distributed throughout the day.
- The highest event frequencies are observed during:
  - 21:00 Hours (817 events)
  - 20:00 Hours (685 events)
  - 06:00 Hours (662 events)
  - 05:00 Hours (661 events)
- Event frequency declines significantly during midday hours:
  - 13:00 Hours (33 events)
  - 14:00 Hours (15 events)
  - 15:00 Hours (9 events)
  - 16:00 Hours (9 events)
- A secondary increase in event volume is observed during late evening hours.

---

## Insight

- Traffic incidents and operational events exhibit strong temporal patterns.
- Early morning and late evening periods appear to be associated with higher event activity.
- Midday hours show relatively low event occurrence, indicating reduced operational disruptions during these periods.
- Time of occurrence is likely to be an important predictive feature for operational priority assessment.

---

## Action

- Create time-based features such as:
  - Hour of Day
  - Peak Hour Indicator
  - Day of Week
  - Weekend Indicator
- Investigate whether high-priority events are concentrated during specific time periods.
- Compare event causes across different hours to identify temporal patterns in traffic disruptions.

In [104]:
df['day_of_week'].value_counts()

,count
day_of_week,
Thursday,1359
Friday,1258
Tuesday,1252
Saturday,1243
Wednesday,1174
Sunday,967
Monday,920


# Day of Week Analysis

## Observation

- Event occurrences are distributed across all days of the week.
- The highest event counts are observed on:
  - Thursday (1,359 events)
  - Friday (1,258 events)
  - Tuesday (1,252 events)
  - Saturday (1,243 events)
- The lowest event counts are observed on:
  - Monday (920 events)
  - Sunday (967 events)

---

## Insight

- Event activity remains relatively consistent throughout the week, indicating that traffic disruptions are not confined to specific weekdays.
- Slightly higher activity during mid-week and late-week periods may reflect increased traffic volumes, operational activities, or infrastructure-related interventions.
- Day of week may contribute useful temporal information for priority prediction and resource planning.

---

## Action

- Retain day-of-week information as a predictive feature.
- Create a Weekend Indicator feature.
- Analyze whether priority levels vary significantly across weekdays and weekends.

In [105]:
df['status'].value_counts()

,count
status,
closed,7095
active,1007
resolved,71


In [106]:
df['police_station'].nunique()

54

In [107]:
df['zone'].value_counts(dropna=False).head(20)

,count
zone,
NaN,4729
Central Zone 2,623
West Zone 1,433
North Zone 2,413
West Zone 2,358
South Zone 2,354
North Zone 1,318
Central Zone 1,269
East Zone 1,253


# Operational and Geographical Attribute Analysis

## Observation

### Status Distribution

| Status | Count |
|----------|--------:|
| Closed | 7,095 |
| Active | 1,007 |
| Resolved | 71 |

- The majority of events have already been closed.
- Active events account for a relatively small portion of the dataset.

### Police Station Distribution

- The dataset contains records from **54 unique police stations**.
- Events are geographically distributed across multiple traffic management jurisdictions.

### Zone Distribution

| Zone | Count |
|--------|--------:|
| NaN | 4,729 |
| Central Zone 2 | 623 |
| West Zone 1 | 433 |
| North Zone 2 | 413 |
| West Zone 2 | 358 |
| South Zone 2 | 354 |
| North Zone 1 | 318 |
| Central Zone 1 | 269 |
| East Zone 1 | 253 |
| South Zone 1 | 233 |
| East Zone 2 | 190 |

- Approximately 58% of records do not contain zone information.
- Available records are distributed across multiple administrative zones, with Central Zone 2 containing the highest number of events.

---

## Insight

- The `status` feature is heavily skewed toward closed events and is likely generated after operational decisions have already been made.
- Since event status reflects outcomes rather than initial event characteristics, it may introduce information leakage if used directly for prediction.
- The presence of 54 police stations indicates substantial spatial diversity within the dataset and suggests that location-specific traffic patterns may exist.
- Although the `zone` attribute contains significant missing values, the available information may still provide useful geographic context for hotspot analysis and visualization.

---

## Action

### Status

- Exclude `status` from model training to prevent target leakage.
- Retain for operational reporting and dashboard analytics.

### Police Station

- Retain as a candidate predictive feature.
- Analyze whether specific police stations experience higher concentrations of high-priority events.

### Zone

- Retain temporarily for exploratory analysis and visualization.
- Evaluate imputation strategies or grouping approaches before deciding whether to include it in the final model.

### Next Steps

- Analyze spatial patterns using:
  - Police Station
  - Zone
  - Latitude
  - Longitude
- Identify geographic hotspots associated with high-priority incidents.
- Assess whether location-based attributes improve predictive performance.

In [108]:
df['corridor'].value_counts().head(20)

,count
corridor,
Non-corridor,3124
Mysore Road,743
Bellary Road 1,610
Tumkur Road,458
Bellary Road 2,379
Hosur Road,298
ORR North 1,275
Old Madras Road,263
Magadi Road,245


In [109]:
df['veh_type'].value_counts(dropna=False)

,count
veh_type,
NaN,3286
bmtc_bus,1466
heavy_vehicle,965
lcv,678
others,449
private_bus,359
private_car,345
truck,276
ksrtc_bus,217


# Traffic Corridor and Vehicle Type Analysis

## Observation

### Corridor Distribution

- Events are distributed across multiple major traffic corridors.
- The most affected corridors include:

| Corridor | Count |
|----------|--------:|
| Non-corridor | 3,124 |
| Mysore Road | 743 |
| Bellary Road 1 | 610 |
| Tumkur Road | 458 |
| Bellary Road 2 | 379 |
| Hosur Road | 298 |
| ORR North 1 | 275 |
| Old Madras Road | 263 |
| Magadi Road | 245 |
| ORR East 1 | 244 |

- A substantial number of incidents occur outside designated corridors.

### Vehicle Type Distribution

| Vehicle Type | Count |
|-------------|--------:|
| Missing | 3,286 |
| BMTC Bus | 1,466 |
| Heavy Vehicle | 965 |
| LCV | 678 |
| Others | 449 |
| Private Bus | 359 |
| Private Car | 345 |
| Truck | 276 |
| KSRTC Bus | 217 |
| Taxi | 95 |
| Auto | 37 |

- Vehicle information is available for approximately 60% of the dataset.
- Public transport and heavy vehicles constitute a significant portion of recorded incidents.

---

## Insight

### Corridor

- Traffic corridors appear to be important operational regions within the transportation network.
- Certain corridors consistently experience higher event frequencies, indicating potential congestion hotspots.
- Corridor information can provide valuable geographic context beyond latitude and longitude coordinates.

### Vehicle Type

- Incidents involving buses, heavy vehicles, and commercial vehicles are common.
- Larger vehicles are more likely to cause lane obstruction, traffic disruption, and operational challenges.
- Despite moderate missingness, vehicle type may contribute useful predictive information for priority assessment.

---

## Action

### Corridor

- Retain `corridor` as a candidate feature for modeling.
- Use corridor-level aggregation for hotspot identification and dashboard visualizations.
- Analyze the relationship between corridor and event priority.

### Vehicle Type

- Retain `veh_type` for further evaluation.
- Create a separate category for missing vehicle information if required.
- Investigate whether specific vehicle categories are associated with higher operational priority.

### Next Steps

- Examine the relationship between:
  - Corridor and Priority
  - Vehicle Type and Priority
- Identify whether specific corridors or vehicle categories consistently generate high-priority incidents.

In [110]:
pd.crosstab(df['corridor'], df['priority']).sort_values('High', ascending=False).head(15)

priority,High,Low
corridor,,
Mysore Road,741,2
Bellary Road 1,610,0
Tumkur Road,454,4
Bellary Road 2,379,0
Hosur Road,298,0
ORR North 1,275,0
Old Madras Road,263,0
Magadi Road,245,0
ORR East 1,244,0


# Corridor vs Priority Analysis

## Observation

- Several major corridors exhibit an overwhelmingly high proportion of high-priority events.
- Some corridors contain almost exclusively high-priority incidents.
- Examples include:
  - Bellary Road 1
  - Bellary Road 2
  - Hosur Road
  - ORR North 1
  - Old Madras Road

---

## Insight

- Corridor information appears to have a strong relationship with operational priority.
- The observed pattern is unusually strong and requires further investigation to determine whether it reflects genuine traffic behavior or a data collection process.
- There is a possibility that corridor assignment itself is correlated with operational decision-making, which may introduce information leakage.

---

## Action

- Validate whether corridor information is available at the time of event reporting.
- Investigate the relationship between corridor and event type.
- Assess whether corridor should be used as a predictive feature or reserved for operational analytics and visualization.

In [111]:
pd.crosstab(
    df['corridor'],
    df['event_type']
).head(20)

event_type,planned,unplanned
corridor,,
Airport New South Road,17,50
Bannerghata Road,5,204
Bellary Road 1,11,599
Bellary Road 2,14,365
CBD 1,1,25
CBD 2,6,98
Hennur Main Road,5,91
Hosur Road,2,296
IRR(Thanisandra road),4,91


In [112]:
df.groupby('corridor')['priority'].value_counts(normalize=True)

,,proportion
corridor,priority,
Airport New South Road,High,1.000000
Bannerghata Road,High,1.000000
Bellary Road 1,High,1.000000
Bellary Road 2,High,1.000000
CBD 1,High,1.000000
CBD 2,High,1.000000
Hennur Main Road,High,1.000000
Hosur Road,High,1.000000
IRR(Thanisandra road),High,1.000000


# Corridor vs Priority Analysis

## Observation

- A remarkably strong relationship exists between corridor and priority.
- Most designated traffic corridors contain almost exclusively high-priority events.
- Non-corridor locations contain almost exclusively low-priority events.

Examples:

| Corridor | High Priority Proportion |
|-----------|-------------------------:|
| Bellary Road 1 | 100% |
| Bellary Road 2 | 100% |
| Hosur Road | 100% |
| ORR North 1 | 100% |
| Old Madras Road | 100% |
| Mysore Road | 99.7% |
| Tumkur Road | 99.1% |

In contrast:

| Corridor | Low Priority Proportion |
|-----------|------------------------:|
| Non-corridor | 100% |

---

## Insight

- Corridor demonstrates an exceptionally strong association with operational priority.
- The observed pattern is unlikely to arise solely from natural traffic conditions.
- This suggests that corridor assignment may be influenced by administrative or operational policies.
- There is a high risk of target leakage if corridor is directly used for priority prediction.

---

## Action

- Exclude corridor from the primary priority prediction model.
- Retain corridor for:
  - Geographic hotspot analysis
  - Dashboard visualizations
  - Resource planning insights
- Focus model development on features available before priority assignment, such as:
  - Event Cause
  - Event Type
  - Time Features
  - Location Coordinates
  - Vehicle Type

In [113]:
pd.crosstab(df['veh_type'], df['priority'])

priority,High,Low
veh_type,,
auto,25,12
bmtc_bus,876,590
heavy_vehicle,623,340
ksrtc_bus,153,64
lcv,477,201
others,326,123
private_bus,229,130
private_car,268,77
taxi,78,17


In [114]:
pd.crosstab(df['event_type'], df['priority'])

priority,High,Low
event_type,,
planned,275,192
unplanned,4755,2949


# Vehicle Type and Event Type vs Priority Analysis

## Observation

### Vehicle Type vs Priority

| Vehicle Type | High Priority | Low Priority |
|-------------|--------------:|-------------:|
| BMTC Bus | 876 | 590 |
| Heavy Vehicle | 623 | 340 |
| LCV | 477 | 201 |
| Others | 326 | 123 |
| Private Car | 268 | 77 |
| Private Bus | 229 | 130 |
| Truck | 181 | 95 |
| KSRTC Bus | 153 | 64 |
| Taxi | 78 | 17 |
| Auto | 25 | 12 |

### Event Type vs Priority

| Event Type | High Priority | Low Priority |
|------------|--------------:|-------------:|
| Planned | 275 | 192 |
| Unplanned | 4755 | 2949 |

- High-priority incidents outnumber low-priority incidents across most vehicle categories.
- Public transport and commercial vehicles contribute significantly to recorded disruptions.
- Unplanned events constitute the majority of both high-priority and low-priority incidents.

---

## Insight

### Vehicle Type

- Vehicle type appears to influence operational priority.
- Larger vehicles such as buses, trucks, and heavy vehicles are more likely to cause significant traffic disruption due to lane blockage and reduced road capacity.
- Vehicle type may provide useful predictive information when combined with event cause and location.

### Event Type

- Event type alone does not strongly determine operational priority.
- Both planned and unplanned events contain substantial proportions of high-priority incidents.
- Additional contextual features such as event cause, location, and time are required to accurately assess event severity.

---

## Action

- Retain both `veh_type` and `event_type` as candidate predictive features.
- Create a separate category for missing vehicle information.
- Evaluate feature importance during model training.
- Combine vehicle, temporal, and event-related features to improve priority prediction performance.

In [115]:
pd.crosstab(
    df['event_type'],
    df['priority'],
    normalize='index'
).round(3) * 100

priority,High,Low
event_type,,
planned,58.9,41.1
unplanned,61.7,38.3


# Event Type vs Priority Analysis

## Observation

| Event Type | High Priority (%) | Low Priority (%) |
|------------|------------------:|-----------------:|
| Planned | 58.9 | 41.1 |
| Unplanned | 61.7 | 38.3 |

- Both planned and unplanned events exhibit similar priority distributions.
- High-priority incidents dominate in both categories.
- The difference between planned and unplanned events is relatively small.

---

## Insight

- Event type alone is not a strong discriminator of operational priority.
- Planned events are not necessarily less severe than unplanned events.
- Additional contextual factors such as event cause, location, vehicle type, and temporal attributes are likely more influential in determining event severity.

---

## Action

- Retain `event_type` as a supporting feature rather than a primary predictor.
- Focus feature engineering efforts on higher-impact attributes such as:
  - Event Cause
  - Vehicle Type
  - Time Features
  - Geographic Information
- Evaluate feature importance during model training to quantify the contribution of event type.

##Phase 3: Data Cleaning

In [116]:
df_clean = df.copy()

print("Original Shape:", df.shape)
print("Working Shape:", df_clean.shape)

Original Shape: (8173, 49)
Working Shape: (8173, 49)


In [117]:
drop_useless = [
    'comment',
    'map_file',
    'meta_data',
    'direction',
    'route_path',
    'assigned_to_police_id',
    'citizen_accident_id',
    'cargo_material',
    'reason_breakdown',
    'age_of_truck'
]

df_clean.drop(columns=drop_useless, inplace=True)

In [118]:
drop_leakage = [
    'status',
    'closed_by_id',
    'closed_datetime',
    'resolved_by_id',
    'resolved_datetime',
    'resolved_at_address',
    'resolved_at_latitude',
    'resolved_at_longitude'
]

df_clean.drop(columns=drop_leakage, inplace=True)

In [119]:
drop_ids = [
    'id',
    'created_by_id',
    'last_modified_by_id',
    'kgid',
    'gba_identifier',
    'client_id'
]

df_clean.drop(columns=drop_ids, inplace=True)

In [120]:
print("Shape after cleaning:", df_clean.shape)

df_clean.info()

Shape after cleaning: (8173, 25)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8173 entries, 0 to 8172
Data columns (total 25 columns):
 #   Column                 Non-Null Count  Dtype              
---  ------                 --------------  -----              
 0   event_type             8173 non-null   object             
 1   latitude               8173 non-null   float64            
 2   longitude              8173 non-null   float64            
 3   endlatitude            8004 non-null   float64            
 4   endlongitude           8004 non-null   float64            
 5   address                8170 non-null   object             
 6   end_address            687 non-null    object             
 7   event_cause            8173 non-null   object             
 8   requires_road_closure  8173 non-null   bool               
 9   start_datetime         8173 non-null   datetime64[ns, UTC]
 10  end_datetime           490 non-null    object             
 11  authenticated          

# Data Cleaning

## Observation

- Several columns contained extremely high missing values (>95%).
- Multiple attributes represented administrative identifiers rather than meaningful event characteristics.
- Some features contained information generated after event handling, creating a risk of target leakage.

---

## Insight

- Retaining highly sparse and leakage-prone features may negatively impact model reliability and interpretability.
- Removing irrelevant operational metadata helps focus the model on information available at the time of event occurrence.
- Data cleaning improves feature quality while reducing dimensionality.

---

## Action

- Removed highly sparse columns.
- Removed administrative identifier columns.
- Removed leakage-prone attributes that contain post-event information.
- Retained operational, temporal, vehicle, and geographic features for subsequent feature engineering and modeling.

In [121]:
missing = (
    df_clean.isnull()
    .sum()
    .sort_values(ascending=False)
)

missing = missing[missing > 0]

missing

,0
end_datetime,7683
end_address,7486
junction,5663
zone,4729
veh_no,3287
veh_type,3286
description,1360
endlatitude,169
endlongitude,169
corridor,20


In [123]:
print("Rows before:", df_clean.shape[0])
df_clean = df_clean.dropna(subset=['priority'])
print("Rows after:", df_clean.shape[0])

Rows before: 8171
Rows after: 8171


## Missing Target Value Handling

### Observation

- The `priority` column contained 2 missing values.
- Since `priority` serves as the target variable for model training, records without a target label cannot be used for supervised learning.

---

### Insight

- Retaining unlabeled records would not contribute to model training and could introduce inconsistencies during evaluation.
- The number of missing target values is negligible (<0.03% of the dataset).

---

### Action

- Removed records with missing `priority` values.
- Retained only fully labeled observations for subsequent modeling and evaluation.

In [124]:
df_clean.to_csv(
    "cleaned_traffic_events.csv",
    index=False
)

print("Dataset saved successfully.")

Dataset saved successfully.
